lpc55_sign/src/signed_image.rs

```diff
@@ -24,366 +36,166 @@ fn get_pad(val: usize) -> usize {
     }
 }
 
-#[derive(Clone, Debug, Deserialize)]
-#[serde(rename_all = "kebab-case", deny_unknown_fields)]
-pub struct CertChain {
-    pub cert_paths: Vec<PathBuf>,
-    pub priv_key: Option<PathBuf>,
-}
-
-#[derive(Clone, Debug, Deserialize)]
-#[serde(rename_all = "kebab-case", deny_unknown_fields)]
-pub struct CfgFile {
-    pub certs: Vec<CertChain>,
-}
-
-fn validate_certs(certs: &[CertChain]) -> Result<()> {
-    if certs.len() > 4 {
-        return Err(anyhow!("Too many certificate chains, max is 4"));
-    }
-
-    let cnt = certs.iter().filter(|c| c.priv_key.is_some()).count();
-
-    if cnt != 1 {
-        return Err(anyhow!(
-            "Exactly one certificate chain should specify a private key for signing"
-        ));
+fn pad_roots(mut roots: Vec<Vec<u8>>) -> Result<Vec<Vec<u8>>> {
+    match roots.len().cmp(&4) {
+        Ordering::Equal => Ok(roots),
+        Ordering::Less => {
+            let empty = [vec![], vec![], vec![], vec![]];
+            roots.extend_from_slice(&empty[..4 - roots.len()]);
+            Ok(roots)
+        }
+        Ordering::Greater => bail!("Too many roots, max four"),
     }
-
-    Ok(())
 }
 
-pub fn sign_chain(
-    binary_path: &Path,
-    cert_path_prefix: Option<&Path>,
-    certs: &[CertChain],
-    outfile_path: &Path,
+/// Prepare an image for signing: append a certificate table,
+/// write the augmented length into the header, and compute &
+/// append the root-key-hash table. Returns the stamped image
+/// and the root-key-table hash.
+pub fn stamp_image(
+    mut image_bytes: Vec<u8>,
+    signing_certs: Vec<Vec<u8>>,
```

Comment:
Could we pass around `&[Vec<u8>]` instead, to reduce allocation?

Reply:
I found references to vectors difficult to manipulate when I first tried this. I switched to passing ownership and not worrying about allocation because (a) I could get it to work without fighting the borrow checker all the time, and (b) because this is running on the host, not the target system, I don't think that copying a few hundred bytes unnecessarily costs very much.

Reply:
👍🏻